From d7a238a018cbb1d8a4862f6ccbeae77bf5a4959f Mon Sep 17 00:00:00 2001
From: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date: Tue, 6 Jan 2015 15:30:20 -0500
Subject: [input] alps: fix v4 button press recognition

Message-id: <1420558220-2988-1-git-send-email-benjamin.tissoires@redhat.com>
Patchwork-id: 102493
O-Subject: [RHEL7.1 PATCH 5/4] Input: alps - fix v4 button press recognition
Bugzilla: 1107819
RH-Acked-by: Aristeu S. Rozanski F. <arozansk@redhat.com>
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1107819
Upstream: since v3.18
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8459662
Tested by me on a Toshiba Tecra W50-A with a v7 Alps touchpad

The original commit message is quite misleading.
The problem we have here is that for the value stored for the right button
when it is pressed is an int, and will be 2 (packet[4] & 0x02).

Sending 2 as an input event means that the button is auto-repeated by the kernel
and is ignored by Xorg. Forcing the value to 1 fixes that.

commit b0cfb794a3dd1d699f3e453f9180bd06508fb8f0
Author: Andreas Bosch <linux@progandy.de>
Date:   Wed Oct 15 10:44:50 2014 -0700

    Input: alps - fix v4 button press recognition

    Since the change to struct input_mt_pos some variables are now bitfields
    instead of integers. Automatic conversion from integer to bitfield entry
    destroys information, therefore enforce boolean interpretation instead.

    Link: https://bugzilla.redhat.com/show_bug.cgi?id=1114768
    Fixes: 02d04254a5df ("Input: alps - use struct input_mt_pos to track coordinates")
    Signed-off-by: Andreas Bosch <linux@progandy.de>
    Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/input/mouse/alps.c b/drivers/input/mouse/alps.c
index e48ce93..f300780 100644
--- a/drivers/input/mouse/alps.c
+++ b/drivers/input/mouse/alps.c
@@ -835,8 +835,8 @@ static void alps_process_packet_v4(struct psmouse *psmouse)
   f->fingers = alps_process_bitmap(priv, f);
  }
 
- f->left = packet[4] & 0x01;
- f->right = packet[4] & 0x02;
+ f->left = !!(packet[4] & 0x01);
+ f->right = !!(packet[4] & 0x02);
 
  f->st.x = ((packet[1] & 0x7f) << 4) | ((packet[3] & 0x30) >> 2) |
     ((packet[0] & 0x30) >> 4);
-- 
1.7.1